presto-docs/src/main/sphinx/sql/select.rst

```diff
@@ -92,6 +92,95 @@ Additionally, the relations within a ``WITH`` clause can chain::
     relation is used. This means that if the relation is used more than once and the query
     is non-deterministic, the results may be different each time.
 
+WITH RECURSIVE Clause
+---------------------
+
+The ``WITH RECURSIVE`` clause is a variant of the ``WITH`` clause. It can define
+a list of queries to process, including recursive processing of suitable
+queries.
+
+.. warning::
+
+    This feature is experimental only. Proceed to use it only if you understand
+    potential query failures and the impact of the recursion processing on your
+    workload.
+
+.. code-block:: none
+
+   WITH RECURSIVE T(n) AS (
+    <recursion base relation>
+    UNION (ALL | DISTINCT)
+    <recursion step relation: SELECT  FROM ... T...>
+    )
+    ... SELECT ... FROM T
+
+A recursive ``WITH``-query must be shaped as a ``UNON`` of two relations. The
+first relation is called the *recursion base*, and the second relation is called
+the *recursion step*. Presto supports recursive ``WITH``-queries with a single
+recursive reference to a ``WITH``-query from within the query. The name ``T`` of
+the query ``T`` can be mentioned once in the ``FROM`` clause of the recursion
+step relation.
+
+The following listing shows a simple example, that displays a commonly used
+form of only one query in the list:
+
+.. code-block:: none
+
+    WITH RECURSIVE t(n) AS (
+        VALUES (1)
+        UNION ALL
+        SELECT n+1 FROM t WHERE n < 4
+    )
+    SELECT sum(n) FROM t;
+
+In the preceding query the simple assignment ``VALUES (1)`` defines the
+recursion base relation. ``SELECT n+1 FROM t WHERE n < 4`` defines the recursion
+step relation. The recursion processing performs these steps:
+
+- first recursion yields ``1 + 1 = 2``
+- second recursion uses the result from the first and adds one: ``2 + 1 = 3``
+- third recursion users the result from the second and adds one again:
+  ``3 + 1 = 4``
+- fourth recursion aborts since ``n = 4``
+- this results in ``t`` having values ``2``, ``3`` and ``4``
+- the final statement performs the sum operation of these elements with the
+  final result value ``10``
```

Comment:
To simplify the example, I suggest that we do just `SELECT * FROM t` in the final line instead of `sum(n)`.
That would clearly show that the resulting table consists of all the intermediate results including the recursion base. It should have one column with rows `1`, `2`, `3`, `4`.

Reply:
Then we loose the syntax that exposes the n outside as well. Its simple enough to understand now, yet more powerful for users. I will leave it as it stands.